File: sdk/dart/symbol_sdk/lib/symbol/idGenerator.dart
import 'dart:typed_data';
import './Network.dart';
import '../utils/converter.dart';
import 'package:pointycastle/export.dart';

final NAMESPACE_FLAG = 1 << 63;

int generateMosaicId(SymbolAddress ownerAddress, int nonce) {
  final hasher = SHA3Digest(256);
  hasher.update(intToBytes(nonce, 4), 0, 4);
  hasher.update(ownerAddress.bytes, 0, ownerAddress.bytes.length);
  var digest = Uint8List(32);
  hasher.doFinal(digest, 0);

  var result = bytesToInt(digest, 8);
  if (result & NAMESPACE_FLAG != 0){
    result -= NAMESPACE_FLAG;
  }
  return result;
}

int generateNamespaceId(dynamic name, [int parentNamespaceId = 0]) {
  if (name is String) {
    name = Uint8List.fromList(name.codeUnits);
  }
  final hasher = SHA3Digest(256);
  hasher.update(intToBytes((parentNamespaceId & 0xFFFFFFFF).toInt(), 4), 0, 4);
  hasher.update(intToBytes(((parentNamespaceId >> 32) & 0xFFFFFFFF).toInt(), 4), 0, 4);
  hasher.update(name, 0, name.length);
  var digest = Uint8List(32);
  hasher.doFinal(digest, 0);
  print(bytesToHex(digest));
  
  final result = bytesToInt(digest, 8);
  return result | NAMESPACE_FLAG;
}

bool isValidNamespaceName(String name) {
  bool isAlphanum(String character) => 
    ('a'.codeUnitAt(0) <= character.codeUnitAt(0) && character.codeUnitAt(0) <= 'z'.codeUnitAt(0)) || 
    ('0'.codeUnitAt(0) <= character.codeUnitAt(0) && character.codeUnitAt(0) <= '9'.codeUnitAt(0));
  
  if (name.isEmpty || !isAlphanum(name[0])){
    return false;
  }

  for (var i = 0; i < name.length; i++) {
    final ch = name[i];
    if (!isAlphanum(ch) && '_' != ch && '-' != ch){
      return false;
    }
  }

  return true;
}

List<int> generateNamespacePath(String fullyQualifiedName) {
  final path = <int>[];
  var parentNamespaceId = 0;
  for (var name in fullyQualifiedName.split('.')) {
    if (!isValidNamespaceName(name)){
      throw ArgumentError('fully qualified name is invalid due to invalid part name ($fullyQualifiedName)');
    }

    path.add(generateNamespaceId(name, parentNamespaceId));
    parentNamespaceId = path.last;
  }

  return path;
}

File: sdk/php/src/symbol/idGenerator.php
<?php

namespace SymbolSdk\Symbol;

use Exception;

class IdGenerator {
  private const NAMESPACE_FLAG = 0x8000000000000000;

  public static function generateMosaicId($ownerAddress, $nonce) {
    $hasher = hash_init('sha3-256');
    hash_update($hasher, self::uint32ToBytes($nonce));
    hash_update($hasher, $ownerAddress->binaryData);
    $digest = hash_final($hasher, true);
    $result = self::digestToGmp($digest);
    
    $gmpMax = gmp_init('9223372036854775807');
    
    if (gmp_testbit($result, 63)) {
        $result = gmp_sub($result, gmp_add($gmpMax, 1));
    }
    
    return gmp_strval($result);
  }

  public static function generateNamespaceId($name, $parentNamespaceId = 0) {
    $hasher = hash_init('sha3-256');
    hash_update($hasher, self::uint32ToBytes($parentNamespaceId & 0xFFFFFFFF));
    hash_update($hasher, self::uint32ToBytes(($parentNamespaceId >> 32) & 0xFFFFFFFF));
    hash_update($hasher, $name);
    $digest = hash_final($hasher, true);
    $result = self::digestToGmp($digest);
    
    return gmp_strval(gmp_or($result, NAMESPACE_FLAG));
  }

  public static function isValidNamespaceName($name) {
    $isAlphanum = function($char) {
        return ctype_alnum($char);
    };

    if (!$name || !$isAlphanum($name[0])) {
        return false;
    }

    for ($i = 0; $i < strlen($name); ++$i) {
        $ch = $name[$i];
        if (!$isAlphanum($ch) && $ch !== '_' && $ch !== '-') {
            return false;
        }
    }

    return true;
  }

  public static function generateNamespacePath($fullyQualifiedName) {
    $path = [];
    $parentNamespaceId = 0;
    
    foreach (explode('.', $fullyQualifiedName) as $name) {
      if (!self::isValidNamespaceName($name)) {
          throw new Exception("fully qualified name is invalid due to invalid part name ({$fullyQualifiedName})");
      }
      
      $path[] = self::generateNamespaceId($name, $parentNamespaceId);
      $parentNamespaceId = end($path);
    }
    
    return $path;
  }

  public static function generateMosaicAliasId($fullyQualifiedName) {
    $path = self::generateNamespacePath($fullyQualifiedName);
    return end($path);
  }

  private static function uint32ToBytes($num) {
    return pack('V', $num);
  }

  private static function digestToGmp($digest) {
    $result = gmp_init(0);
    for ($i = 0; $i < 8; ++$i) {
        $shifted = gmp_mul(ord($digest[$i]), gmp_pow(2, 8 * $i));
        $result = gmp_add($result, $shifted);
    }
    return $result;
  }
}
